symphony-ws/src/main/java/se/havochvatten/symphony/service/ReportService.java

```diff
@@ -420,4 +415,290 @@ static void writeComparisonRows(Locale locale, SymphonyBandDto[] meta, int[] fea
+            throws SymphonyStandardAppException {
+
+        CompoundComparisonExport result = new CompoundComparisonExport();
+
+        BaselineVersion baseline = em.createNamedQuery("BaselineVersion.getById", BaselineVersion.class)
+            .setParameter("id", comparison.getBaseline().getId())
+            .getSingleResult();
+
+        Map<Integer, String>
+            ecoTitles = metaDataService.getComponentTitles(
+                comparison.getBaseline().getId(),
+                LayerType.ECOSYSTEM, preferredLanguage),
+            pressureTitles = metaDataService.getComponentTitles(
+                comparison.getBaseline().getId(),
+                LayerType.PRESSURE, preferredLanguage);
+
+        ComparisonResult[] cmpResults = excludeZeroes ?
+            comparison.getResult().values().stream()
+                .filter(c -> c.cumulativeTotal != 0)
+                .toArray(ComparisonResult[]::new) :
+            comparison.getResult().values().toArray(new ComparisonResult[0]);
+
+        MultiComparisonAuxiliary[] layersToList = new MultiComparisonAuxiliary[cmpResults.length];
+
+        for (int i = 0; i < cmpResults.length; ++i) {
+            layersToList[i] = getLayerIndicesToListForResult(cmpResults[i], excludeZeroes);
+        }
+
+        result.id = comparison.getId();
+        result.baselineName = baseline.getName();
+        result.name = comparison.getName();
+        result.result = IntStream.range(0, cmpResults.length)
+            .mapToObj(i -> {
+                ComparisonResultExport cmpResult = new ComparisonResultExport();
+                cmpResult.calculationName = cmpResults[i].calculationName;
+                cmpResult.ecosystemTitles = Arrays.stream(layersToList[i].ecosystems)
+                    .mapToObj(e -> ecoTitles.get(cmpResults[i].includedEcosystems[e]))
+                    .toArray(String[]::new);
+                cmpResult.pressureTitles = Arrays.stream(layersToList[i].pressures)
+                    .mapToObj(p -> pressureTitles.get(cmpResults[i].includedPressures[p]))
+                    .toArray(String[]::new);
+                cmpResult.comparisonMatrix = layersToList[i].result;
+                cmpResult.cumulativeTotal = cmpResults[i].cumulativeTotal;
+                return cmpResult;
+
+            }).toArray(ComparisonResultExport[]::new);
+
+        return result;
+    }
+
+    private OdfTable createSheet(OdfSpreadsheetDocument doc, int rows, int cols, String title) {
+        OdfTable newSheet = OdfTable.newTable(doc, rows, Math.min(cols, 5));
+
+        newSheet.getCellRangeByPosition(2, 1, 5, 1).merge();
+        OdfTableCell titleCell = newSheet.getCellByPosition(2, 1);
+        titleCell.setStringValue(title);
+        titleCell.getOdfElement().setStyleName("cmpName");
+
+        newSheet.getColumnByIndex(0).setWidth(5);
+        newSheet.getRowByIndex(1).setHeight(14, false);
+
+        return newSheet;
+    }
+
+    public byte[] generateMultiComparisonAsODS(
```

Comment:
This is a very long function and could maybe have extracted snippets into separate functions. Since it is urgent to get this out into 1.17 maybe it's not super important right now, but could make it more readable.

Reply:
It is a long procedure, but I've been reluctant to splitting it up, mainly to avoid passing around large object values. There are also some opaque quirks and side effects of the utilized library ("odftoolkit") that plays a role in that decision.

However, I do feel there is some regrettable repetition that can probably be mitigated by introducing a couple of utility functions. I'll make a note of it as a `TODO` so that this insight hopefully won't be forgotten.